### PR TITLE
[API-1250] Rename `cloneWithBuilder` to `newBuilderWithClone` for `GenericRecord`s

### PR DESCRIPTION
**Pull request by mdumandag**
Also, the APIs of the `GenericRecord` and `GenericRecordBuilder`
interfaces are updated to reflect this change, along with formatting
fixes.

Also, lots of the documentation is corrected to have the proper
wording for Compact serialization and some code snippets are updated
with the new APIs.

### CODE REVIEW

**Review by ihsandemir — APPROVED** [review]

**Comment by srknzl** [comment]
@mdumandag  Can you also rename the following? 
image

File: hazelcast/src/main/java/com/hazelcast/nio/serialization/GenericRecord.java [hunk 1]
```diff
@@ -30,87 +30,103 @@
...
     /**
-     * Creates a {@link GenericRecordBuilder} allows to create a new object. This method is a convenience method to get a builder,
-     * without creating the class definition for this type. Here you can see  a  new object is constructed from an existing
-     * GenericRecord with its class definition:
-     *
-     * <pre>
+     * Creates a {@link GenericRecordBuilder} allows to create a new object.
+     * This method is a convenience method to get a builder, without creating
+     * the schema/class definition for this type. Here you can see a new object
+     * is constructed from an existing GenericRecord with its schema:
```
Comment (srknzl) [inline hazelcast/src/main/java/com/hazelcast/nio/serialization/GenericRecord.java:86]:
```suggestion
     * is constructed from an existing GenericRecord with its schema/class definition:
```

File: hazelcast/src/main/java/com/hazelcast/nio/serialization/GenericRecordBuilder.java [hunk 2]
```diff
@@ -68,775 +80,1025 @@ static GenericRecordBuilder compact(String typeName) {
...
      */
     @Nonnull
     GenericRecordBuilder setGenericRecord(@Nonnull String fieldName, @Nullable GenericRecord value);
 
     /**
-     * Sets a decimal which is arbitrary precision and scale floating-point number
-     * It is legal to set the field again only when Builder is created with {@link GenericRecord#cloneWithBuilder()}.
-     * Otherwise, it is illegal to set to the same field twice.
+     * Sets a decimal which is arbitrary precision and scale floating-point
+     * number It is legal to set the field again only when Builder is created
```
Comment (srknzl) [inline hazelcast/src/main/java/com/hazelcast/nio/serialization/GenericRecordBuilder.java:484]:
```suggestion
     * Sets a decimal which is arbitrary precision and scale floating-point
     * number. It is legal to set the field again only when Builder is created
```

We don't have the first sentence in other methods: `Sets a decimal which is arbitrary precision and scale floating-point number.`. I think we should remove these from all of them or add to all of them. for you info, the first sentence only appears in decimal, datetime methods and arrays of them.

File: hazelcast/src/main/java/com/hazelcast/nio/serialization/GenericRecordBuilder.java [hunk 3]
```diff
@@ -68,775 +80,1025 @@ static GenericRecordBuilder compact(String typeName) {
...
-     * Supported only for {@link com.hazelcast.config.CompactSerializationConfig Compact}. Not applicable to {@link Portable}.
+     * Supported only for
+     * {@link com.hazelcast.config.CompactSerializationConfig Compact}. Not
+     * applicable to {@link Portable}.
      *
-     * @param fieldName name of the field as it is defined in its class definition.
-     *                  It should be composed of only alpha-numeric characters.
-     *                  See {@link ClassDefinition} for {@link Portable}
+     * @param fieldName name of the field as it is defined in its class
+     *                  definition.
```
Comment (srknzl) [inline hazelcast/src/main/java/com/hazelcast/nio/serialization/GenericRecordBuilder.java:278]:
```suggestion
     * @param fieldName name of the field as it is defined in its schema.
     *                  
```

Similar for other setNullableXX methods

**Review by srknzl — CHANGES_REQUESTED** [review]

File: hazelcast/src/main/java/com/hazelcast/nio/serialization/GenericRecordBuilder.java (hunk 2 above)
Comment (mdumandag) [reply hazelcast/src/main/java/com/hazelcast/nio/serialization/GenericRecordBuilder.java]: I removed the ones from the arrays, as they don't offer any useful information. Regarding the ones in the decimal and date related ones, I think it is OK to have them as an extra description for those, as the method names and the parameters they have are not directly relatable. I don't see it as a problem to have an extra definition only on some of them

**Review by srknzl — APPROVED** [review]

**Pull request merged**